src/JsonApi/Serializer/CollectionNormalizer.php

```diff
@@ -0,0 +1,207 @@
+        // could be already an array)
+        $returnDataArray = [];
+        if (isset($context['api_sub_level'])) {
+            foreach ($data as $index => $obj) {
+                $returnDataArray['data'][][$index] = $this->normalizer->normalize($obj, $format, $context);
+            }
+
+            return $data;
+        }
+
+        $resourceClass = $this->resourceClassResolver->getResourceClass(
+            $data,
+            $context['resource_class'] ?? null,
+            true
+        );
+
+        $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
+
+        $context = $this->initContext($resourceClass, $context);
+
+        $parsed = IriHelper::parseIri($context['request_uri'] ?? '/', $this->pageParameterName);
+        $paginated = $isPaginator = $data instanceof PaginatorInterface;
+
+        if ($isPaginator) {
+            $currentPage = $data->getCurrentPage();
+            $lastPage = $data->getLastPage();
+            $itemsPerPage = $data->getItemsPerPage();
+
+            $paginated = 1. !== $lastPage;
+        }
+
+        $returnDataArray = [
+            'data' => [],
+            'links' => [
+                'self' => IriHelper::createIri(
+                    $parsed['parts'],
+                    $parsed['parameters'],
+                    $this->pageParameterName,
+                    $paginated ? $currentPage : null
+                ),
+            ],
+        ];
+
+        if ($paginated) {
+            $returnDataArray['links']['first'] = IriHelper::createIri(
+                $parsed['parts'],
+                $parsed['parameters'],
+                $this->pageParameterName,
+                1.
+            );
+
+            $returnDataArray['links']['last'] = IriHelper::createIri(
+                $parsed['parts'],
+                $parsed['parameters'],
+                $this->pageParameterName,
+                $lastPage
+            );
+
+            if (1. !== $currentPage) {
+                $returnDataArray['links']['prev'] = IriHelper::createIri(
+                    $parsed['parts'],
+                    $parsed['parameters'],
+                    $this->pageParameterName,
+                    $currentPage - 1.
+                );
+            }
+
+            if ($currentPage !== $lastPage) {
+                $returnDataArray['links']['next'] = IriHelper::createIri(
+                    $parsed['parts'],
+                    $parsed['parameters'],
+                    $this->pageParameterName,
+                    $currentPage + 1.
+                );
+            }
+        }
+
+        $identifier = null;
+        foreach ($data as $item) {
+            $normalizedItem = $this->normalizer->normalize($item, $format, $context);
+
+            if (!isset($normalizedItem['data'])) {
+                throw new RuntimeException(
+                    'data key expected but not found during JSON API normalization'
```

Comment:
`The 'data' key was expected during JSON API normalization, (but it does not exist or but is not found)`

WDYT ?